### PR TITLE
feat: Check function definitions with inout arguments

### LINKED ISSUE #332: Check function definitions with inout arguments

### PR DESCRIPTION
**Pull request by mark-koch**
Closes #332

Most of the diff are tests, so don't be scared :D

Updates to the checking logic are:

* Update dataflow analysis to mark inout arguments as implicitly used in the exit BB
* The use in the exit BB is a special `InoutReturnSentinel` AST node defined in `nodes.py`
* Update linearity checker to disallow usage of inout varaiables except for `@inout` positions
* Add special error message for fields that are moved out of an inout argument but not moved back before returning

Notes:
* Test validation is disabled until #333
* Shadowing of inout arguments will get a better error message in #338

### CODE REVIEW

File: guppylang/checker/cfg_checker.py [hunk 1]
```diff
@@ -85,7 +86,7 @@ def check_cfg(
     while len(queue) > 0:
         pred, num_output, bb = queue.popleft()
         input_row = [
-            Variable(v.name, v.ty, v.defined_at)
+            Variable(v.name, v.ty, v.defined_at, v.flags)
```
Comment (mark-koch) [inline guppylang/checker/cfg_checker.py:89]: Drive-by fix: The flags should be preserved across BBs

**Review by ss2165 — APPROVED** [review]
LGTM, some small comments

File: guppylang/checker/linearity_checker.py [hunk 2]
```diff
@@ -101,6 +102,7 @@ def check(self, bb: "CheckedBB", is_entry: bool, globals: Globals) -> Scope:
         for var in bb.sig.input_row:
             for place in leaf_places(var):
                 input_scope.assign(place)
+        # TODO: Set used flags in input_scope
```
[the remark this answers is not in the document]
Comment (mark-koch) [reply guppylang/checker/linearity_checker.py:105]: Oops that's a comment I left to remember the fix for #337. Doesn't belong in this PR
